stan/math/prim/mat/fun/qr_thin_Q.hpp

```diff
@@ -0,0 +1,30 @@
+#ifndef STAN_MATH_PRIM_MAT_FUN_QR_THIN_Q_HPP
+#define STAN_MATH_PRIM_MAT_FUN_QR_THIN_Q_HPP
+
+#include <stan/math/prim/arr/err/check_nonzero_size.hpp>
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/scal/err/check_greater_or_equal.hpp>
+#include <Eigen/QR>
+#include <algorithm>
+
+namespace stan {
+namespace math {
+
+template <typename T>
+Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> qr_thin_Q(
+    const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& m) {
+  typedef Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> matrix_t;
+  check_nonzero_size("qr_Q", "m", m);
+  Eigen::HouseholderQR<matrix_t> qr(m.rows(), m.cols());
+  qr.compute(m);
+  const int min_size = std::min(m.rows(), m.cols());
```

Comment:
Why is this the min of rows and columns? Isn't the thinness just a special case when the number of rows is greater than the number of columns?

Reply:
I will look into this more.

Reply:
In the case where the number of columns of the input matrix is greater than the number of rows, if we multiplied by `matrix_t::Identity(m.rows(), m.cols())`, then the resulting Q matrix would have `m.cols()` columns when it should only have `m.rows()` columns. In other words, in that case we could get away with just returning `qr.householderQ()` and not multiplying by the identity matrix, but since we have to loop through and possibly flip the directions anyway, it does not make sense to branch.